File: backend/app/models/discount.py
from __future__ import annotations

from datetime import UTC, datetime
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base

if TYPE_CHECKING:
    from app.models.product import Product


class Discount(Base):
    __tablename__ = "discounts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    discount_type: Mapped[str] = mapped_column(String(20), nullable=False)
    discount_value: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    bundle_quantity: Mapped[int | None] = mapped_column(Integer, nullable=True)
    stack_with_bundle: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    start_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    end_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
        nullable=False,
    )

    product: Mapped["Product"] = relationship(back_populates="discounts")

File: frontend/src/types/discount.ts
export type DiscountType = "PERCENTAGE" | "FIXED_AMOUNT" | "BUNDLE_PRICE";

export interface Discount {
  id: number;
  product_id: number;
  name: string;
  discount_type: DiscountType;
  discount_value: number | string;
  bundle_quantity: number | null;
  stack_with_bundle: boolean;
  start_at: string;
  end_at: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DiscountInput {
  product_id: number;
  name: string;
  discount_type: DiscountType;
  discount_value: number;
  bundle_quantity?: number | null;
  stack_with_bundle?: boolean;
  start_at: string;
  end_at: string;
  is_active: boolean;
}
